### ISSUE TITLE
Option to seed vhosts (and users) from deployment manifest

### ISSUE THREAD

**Issue by MatthiasWinzeler** [pivotal-cf/cf-rabbitmq-release#57]
When CF platform users need a RMQ service, they can obtain it in a self-service manner using the marketplace/broker.

But we also would like to use the RMQ service on a infrastructure level - and for this, we would prefer to configure the needed vhosts and users directly in the deployment manifest so that they're automatically provisioned on deployment time of the BOSH release.

I.e. the cf-mysql release (that is not only used `as a service`, but very often also on infrastructure level, i.e. to back CF components themselves) offers such an option:
https://github.com/cloudfoundry/cf-mysql-release/blob/62b1e35c31ba412c1647e2a7f916c5fa5f89a6dc/jobs/mysql/spec#L127
https://github.com/cloudfoundry/mariadb_ctrl/blob/b2b272a1ef3bca647bd314b8524dcbb1ca3bba8e/mariadb_helper/mariadb_helper.go#L202

Would you welcome such a PR that adds the option to specify such "seeded" vhosts/users to the RMQ release?

How we imagine it:
```
# add a new configuration option:
  rabbitmq-server.seeded_vhosts:
    description: 'Set of vhosts to seed'
    default: {}
    example: |
      - name: vhost1
        username: user1
        password: pw1
        tags: administrator
        permissions: .* .* .* # maybe also split this up to the individual fields conf/read/write?
      - name: vhost2
        username: user2
        password: pw2
        tags: monitoring
        permissions: conf read write
```

These vhosts, users and permissions would be created upon startup (effectively overwriting already existing user attributes/permissions, similar to how the admin user provisioning is done).

This could then be implemented in a couple of lines of shell script in [rabbitmq-server.init.bash](https://github.com/pivotal-cf/cf-rabbitmq-release/blob/v226.1.0/jobs/rabbitmq-server/templates/rabbitmq-server.init.bash#L56,L89), reusing (and probably refactoring) the methods handling the admin/mgmt user setup.

**Comment by michaelklishin** [comment]
You want [definitions import](http://www.rabbitmq.com/management.html). I highly recommend not reinventing them on top of BOSH.

**Comment by MatthiasWinzeler** [comment]
Are you talking about [Load definitions at startup](http://www.rabbitmq.com/management.html#load-definitions)?

If we can specify this in the deployment manifest - which is not possible currently - this could indeed solve our problem.

**Comment by michaelklishin** [comment]
@MatthiasWinzeler yes, that section. I'm pretty sure there is a way to specify arbitrary config settings in Ops Manager, so presumably there should be a way to do that with straight BOSH.

Even if somehow that's not possible or very difficult, making that possible should be the goal instead of reinventing definitions import with BOSH and `rabbitmqctl`.

**Comment by MatthiasWinzeler** [comment]
@michaelklishin You're right, we could set the config entry `load_definitions` using the  [`rabbitmq-server.config`](https://github.com/pivotal-cf/cf-rabbitmq-release/blob/master/jobs/rabbitmq-server/spec#L48) key.

But: `load_definitions` needs a JSON file on the server itself - so we would need to add a mechanism to the BOSH release to write this file to the server.
What do you guys think about this:

```
# Introduce a new configuration item 
rabbitmq-server.load_definitions:
  description: definitions (JSON) to be imported on RabbitMQ startup
  default: null
  example: |
    {
	"users": [{
		"name": "defuser",
		"password": "defpasswd",
		"tags": "administrator"
	}],
	"vhosts": [{
		"name": "from-def"
	}],
	"permissions": [{
		"user": "defuser",
		"vhost": "from-def",
		"configure": ".*",
		"write": ".*",
		"read": ".*"
	}]
}
```

This would then write the contents as a file to the FS and point the RMQ config entry to it.

**Comment by carlo-colombo** [comment]
@MatthiasWinzeler we recently implemented this feature. We expose `load_definitions` property in the `rabbitmq-server`.  Differently from your suggestion the property is not a json serialized as string but directly a structure that will be serialized and loaded using http://www.rabbitmq.com/management.html#load-definitions.

The feature is available from release `v226.6`

 ```
rabbitmq-server.load_definitions:
  description: definitions (JSON) to be imported on RabbitMQ startup
  default: null
  example:
    vhosts:
    -  {name: from-def}
    users:
    - name: defuser
      password: pwd
      tags: administrator
    permissions:
    - user: defuser
      vhost: from-def
      configure: '*.'
      write: '.*'
      read: '.*'

```

**Comment by MatthiasWinzeler** [comment]
Nice to hear - thank you!
